cmd/nerdctl/login.go

```diff
@@ -169,33 +174,119 @@ func GetDefaultAuthConfig(checkCredStore bool, serverAddress string, isDefaultRe
 	return &res, nil
 }
 
-// Code from github.com/cli/cli/command/registry/login.go
-func loginClientSide(ctx context.Context, cmd *cobra.Command, auth types.AuthConfig) (registrytypes.AuthenticateOKBody, error) {
+func loginClientSide(ctx context.Context, cmd *cobra.Command, auth types.AuthConfig) (string, error) {
+	host := auth.ServerAddress
+	if strings.Contains(host, "://") {
+		u, err := url.Parse(host)
+		if err != nil {
+			return "", err
+		}
+		host = u.Host
+	}
 
-	var insecureRegistries []string
-	insecureRegistry, err := cmd.Flags().GetBool("insecure-registry")
+	var dOpts []dockerconfigresolver.Opt
+	insecure, err := cmd.Flags().GetBool("insecure-registry")
+	if err != nil {
+		return "", err
+	}
+	if insecure {
+		logrus.Warnf("skipping verifying HTTPS certs for %q", host)
+		dOpts = append(dOpts, dockerconfigresolver.WithSkipVerifyCerts(true))
+	}
+	hostsDirs, err := cmd.Flags().GetStringSlice("hosts-dir")
 	if err != nil {
-		return registrytypes.AuthenticateOKBody{}, err
+		return "", err
 	}
-	if insecureRegistry {
-		insecureRegistries = append(insecureRegistries, auth.ServerAddress)
+	dOpts = append(dOpts, dockerconfigresolver.WithHostsDirs(hostsDirs))
+
+	authCreds := func(acArg string) (string, string, error) {
+		if acArg == host {
+			if auth.RegistryToken != "" {
+				// Even containerd/CRI does not support RegistryToken as of v1.4.3,
+				// so, nobody is actually using RegistryToken?
```

Comment:
It's not really necessary and the only use of it I know of was for Docker swarm. It solves the problem of sending unscoped credentials to nodes, however, the tokens end up being long lived and passed directly to registries. The better solution is using scoped refresh tokens. It would be good to support scoping somewhere here, although that does not work well with the `docker login` approach.

Reply:
Thanks, the current goal is to just emulate Swarm-less `docker login`